src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright 2018 Inscope Metrics, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.utility;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+
+/**
+ * A regex replacement utility that can also replace tokens not found in the regex.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public final class RegexAndMapReplacer {
+    /**
+     * Replaces all instances of $n (where n is 0-9) with regex match groups, ${var} with regex capture group or variable (from map) 'var'.
+     *
+     * @param matcher matcher to use
+     * @param input input string to match against
+     * @param replace replacement string
+     * @param variables map of variables to include
+     * @return a string with replacement tokens replaced
+     */
+    public static String replaceAll(final Matcher matcher, final String input, final String replace, final Map<String, String> variables) {
+
+        matcher.reset();
```

Comment:
I'm not nuts about the state sharing between this method and the caller.

Reply:
Fixex